modules/nextflow/src/main/groovy/nextflow/processor/ArrayTaskCollector.groovy

```diff
@@ -0,0 +1,135 @@
+ */
+@Slf4j
+@CompileStatic
+class ArrayTaskCollector {
+
+    private TaskProcessor processor
+
+    private ArrayTaskAware executor
+
+    private int arraySize
+
+    private Lock sync = new ReentrantLock()
+
+    private List<TaskRun> array
+
+    private volatile boolean closed
+
+    ArrayTaskCollector(TaskProcessor processor, int arraySize) {
+        if( processor.executor !instanceof ArrayTaskAware )
+            throw new IllegalArgumentException("Executor '${executor.name}' does not support array jobs")
+        this.processor = processor
+        this.executor = (ArrayTaskAware) processor.executor
+        this.arraySize = arraySize
+        this.array = new ArrayList<>(arraySize)
+    }
+
+    /**
+     * Add a task to the current array, and submit the array when it
+     * reaches the desired size.
+     *
+     * @param handler
+     */
+    void submit(TaskRun task) {
+        // submit task directly if the collector is closed
+        if( closed ) {
+            throw new IllegalStateException("Process is already closed")
+        }
+
+        sync.lock()
+        try {
+            // update the array index attribute
+            task.arrayIndex = array.size()
+            // add task to the array array
+            array.add(task)
+
+            // submit array job when the array is ready
+            if( array.size() == arraySize ) {
+                submit0(array)
+                array = new ArrayList<>(arraySize)
+            }
+        }
+        finally {
+            sync.unlock()
+        }
+    }
+
+    /**
+     * Close the collector, submitting any remaining tasks as a partial array job.
+     *
+     * @param process
+     */
+    void close() {
+        sync.lock()
+
+        try {
+            if( array.size() > 0 )
+                submit0(array)
+
+            closed = true
+        }
+        finally {
+            sync.unlock()
+        }
+    }
+
+    protected void submit0(List<TaskRun> array) {
+        // final submit the task array
+        final task = processor.createTaskArray(array)
+        task.script = createTaskArrayScript(array)
+        executor.submit(task)
+    }
+
+    protected String createTaskArrayScript(List<TaskRun> array) {
+        final allLaunchers = new ArrayList<Path>()
+        final script = new StringBuilder()
+        script.append('tasks=()').append('\n')
+        for( TaskRun t0 : array ) {
+            final l0 = t0.workDir.resolve(TaskRun.CMD_RUN)
+            allLaunchers.add(l0)
+            script.append("tasks+=(${Escape.path(l0)})").append('\n')
+        }
+        // the launcher entry point
+        final index = executor.getArrayIndexName()
+        if( index ) {
+            script << '# execute the i-th task\n'
+            script << '/bin/bash ${tasks[' << index << ']}\n'
+        }
+        else {
+            script << '# execute all tasks\n'
+            script << 'for t in ${tasks[@]}; do /bin/bash $t; done\n'
+        }
```

Comment:
This ideally, will handle the execution of tasks in batches. The implementation could be very similar